sopt/src/main/scala/dagr/sopt/cmdline/CommandLineParser.scala

```diff
@@ -0,0 +1,527 @@
+  *     class.
+  * (2) [[parseMainAndClp]] for the same case as (1), but when the main program itself also has arguments, which come
+  *     prior to the name of the command line program.
+  *
+  * To route the arguments to build the appropriate program, the parser searches for a command line program that is a
+  * sub-class of [[Clp]], and creates a [[CommandLineProgramParser]] to parse the args and return an instance of the
+  * [[Clp]].
+  *
+  * Constructor arguments for command line programs (or main class) can be annotated with [[dagr.sopt.arg]] while
+  * constructors themselves can be annotated with [[dagr.sopt.clp]].  The latter may be omitted, but is extremely
+  * useful for grouping related command line programs and having a common description.
+  *
+  * @param commandLineName the name of the base command line.
+  * @param classTag the [[ClassTag]] of [[Clp]] , usually inferred.
+  * @param typeTag the [[TypeTag]] of [[Clp]] , usually inferred.
+  * @tparam Clp the base class for all command line programs to display and parse on the command line.
+  */
+class CommandLineParser[Clp](val commandLineName: String)
+                            (implicit classTag: ClassTag[Clp], typeTag: TypeTag[Clp])
+  extends CommandLineParserStrings {
+
+  private type ClpClass = Class[_ <: Clp]
+  private type ClpGroupClass = Class[_ <: ClpGroup]
+
+  /**
+    * Finds the class of a command line program using the first argument given.
+    *
+    * @param args the arguments, with the first argument used as the command line program name.
+    * @param classes the set of available command line program classes
+    * @return the class of the command lien program, or an error string if either no arguments were given or no
+    *         command line program matched the given argument.
+    */
+  private def parseClpName(args: Array[String], classes: Set[ClpClass]): Either[ClpClass,String] = {
+
+    if (args.length < 1) {
+      Right(MissingClps)
+    }
+    else {
+      val clazzOption: Option[ClpClass] = classes.find(clazz => 0 == args(0).compareTo(clazz.getSimpleName))
+      clazzOption match {
+        case Some(clazz) => Left(clazz)
+        case None        => Right(unknownClpError(classes, args(0)))
+      }
+    }
+  }
+
+  /** When a command does not match any known command, searches for similar commands, using the same method as GIT, and
+    * provide an appropriate error string, perhaps with suggestions.
+    *
+    * @param classes the set of available command line program classes
+    * @param clpName the name of the command line program.
+    * @return the error string.
+    */
+  private def unknownClpError(classes: Set[ClpClass], clpName: String) : String = {
+    val builder = new StringBuilder
+    val distances: mutable.Map[ClpClass, Integer] = new mutable.HashMap[ClpClass, Integer]
+    var bestDistance: Int = Integer.MAX_VALUE
+    var bestN: Int = 0
+    classes.foreach{ clazz =>
```

Comment:
Could some of this be factored out into a `findSimilar(target:String, options:Traversable[String): Seq[String]`?  It feels like we have similar code in several places now.